File: internal/operations/settings_merge.go
package operations

import (
	"encoding/json"
	"reflect"
)

// MergeSettingsSmart performs a deep merge of settings with intelligent array handling.
// Strategy:
// - Nested dicts: Recursive merge
// - Arrays: Union (deduplicate simple types)
// - Scalars: Overlay takes precedence
func MergeSettingsSmart(base, overlay map[string]interface{}) map[string]interface{} {
	result := make(map[string]interface{})
	for k, v := range base {
		result[k] = v
	}

	for key, value := range overlay {
		if existing, ok := result[key]; !ok {
			// New key - add it
			result[key] = value
		} else if isMap(value) && isMap(existing) {
			// Both maps - recursive merge
			result[key] = MergeSettingsSmart(
				existing.(map[string]interface{}),
				value.(map[string]interface{}),
			)
		} else if isSlice(value) && isSlice(existing) {
			// Both slices - merge with deduplication
			result[key] = mergeLists(existing.([]interface{}), value.([]interface{}))
		} else {
			// Scalar or type mismatch - overlay wins
			result[key] = value
		}
	}

	return result
}

func mergeLists(base, overlay []interface{}) []interface{} {
	result := make([]interface{}, len(base))
	copy(result, base)

	for _, item := range overlay {
		// For simple types, deduplicate
		if isSimpleType(item) {
			found := false
			for _, existing := range result {
				if reflect.DeepEqual(existing, item) {
					found = true
					break
				}
			}
			if !found {
				result = append(result, item)
			}
		} else {
			// For complex types, just append
			result = append(result, item)
		}
	}

	return result
}

func isMap(v interface{}) bool {
	_, ok := v.(map[string]interface{})
	return ok
}

func isSlice(v interface{}) bool {
	return reflect.TypeOf(v).Kind() == reflect.Slice
}

func isSimpleType(v interface{}) bool {
	switch v.(type) {
	case string, int, int64, float64, bool, nil:
		return true
	default:
		return false
	}
}

// LoadJSONSettings loads JSON settings from bytes.
func LoadJSONSettings(data []byte) (map[string]interface{}, error) {
	var settings map[string]interface{}
	if err := json.Unmarshal(data, &settings); err != nil {
		return nil, err
	}
	return settings, nil
}

// SaveJSONSettings saves JSON settings to bytes with indentation.
func SaveJSONSettings(settings map[string]interface{}) ([]byte, error) {
	return json.MarshalIndent(settings, "", "  ")
}

File: src/agentctl/operations/settings_merge.py
"""Settings merge utilities for smart JSON merging."""

from __future__ import annotations


def merge_settings_smart(base: dict, overlay: dict) -> dict:
    """Deep merge settings with intelligent array handling.

    Strategy:
    - Nested dicts: Recursive merge
    - Arrays: Union (deduplicate simple types)
    - Scalars: Overlay takes precedence

    Args:
        base: Existing settings
        overlay: New settings to merge in

    Returns:
        Merged settings dictionary
    """
    result = base.copy()

    for key, value in overlay.items():
        if key not in result:
            # New key - add it
            result[key] = value
        elif isinstance(value, dict) and isinstance(result[key], dict):
            # Both dicts - recursive merge
            result[key] = merge_settings_smart(result[key], value)
        elif isinstance(value, list) and isinstance(result[key], list):
            # Both lists - merge with deduplication
            result[key] = _merge_lists(result[key], value)
        else:
            # Scalar or type mismatch - overlay wins
            result[key] = value

    return result


def _merge_lists(base: list, overlay: list) -> list:
    """Merge two lists with deduplication for simple types."""
    result = base.copy()

    for item in overlay:
        # For simple types (str, int, bool), deduplicate
        if isinstance(item, (str, int, bool, float, type(None))):
            if item not in result:
                result.append(item)
        else:
            # For complex types (dict, list), just append
            result.append(item)

    return result
